### ISSUE TITLE
Provide suggestions on how to distribute public keys used to sign attestations

### ISSUE THREAD

**Issue by TomHennen** [slsa-framework/slsa#101]
We should include recommendations on how to convey the maximum SLSA level the keys are valid for.

For example:

* Publish on a website
* Some sort of PKI with an accreditation body signing certs with a different root cert for each SLSA level
* [Sigstore](https://sigstore.dev/) integration
* ...

**Comment by TomHennen** [comment]
Another (maybe bad!) thought:

Let people register a builder id and public key by sending PRs to some GitHub repo.

**Comment by joshuagl** [comment]
Datadog do this for their in-toto links using [TUF](https://theupdateframework.io/), see [Secure Publication of Datadog Agent Integrations with TUF and in-toto](https://www.datadoghq.com/blog/engineering/secure-publication-of-datadog-agent-integrations-with-tuf-and-in-toto/).

I'd be happy to help document how to use TUF for this. 

In fact, I think we could even provide some PoC code. @jku has been experimenting with a tool to store TUF metadata in a GitHub repository, then using GitHub Actions to generate the snapshot and timestamp metadat.

**Comment by trishankatdatadog** [comment]
This is where [the sigstore root of trust](https://blog.sigstore.dev/a-new-kind-of-trust-root-f11eeeed92ef) might be helpful 🙂 @dlorenc @asraa

The idea, as Joshua suggested, is to use TUF to bootstrap trust for in-toto layouts (the functionaries of whom can then use, for example, SPIFFE/SPIRE for machine signing keys). Happy to help with this document, Joshua, if you get it started.

**Comment by TomHennen** [comment]
Oh I think some proposals would be great. Is the idea that everyone that wants to verify DSSE sigs would use in-toto to verify the roots?

**Comment by trishankatdatadog** [comment]
> Oh I think some proposals would be great. Is the idea that everyone that wants to verify DSSE sigs would use in-toto to verify the roots?

No, we mean that people who verify in-toto links (whether using DSSE or not) will probably want to use TUF to verify the in-toto layouts and their associated public keys (not to mention the links and artifacts themselves, but that's a different story for another day) 🙂

**Comment by TomHennen** [comment]
Oh I see.  With this issue I was hoping to find an approach that worked for all users of DSSE (whether in-toto links or not). 😄

**Comment by trishankatdatadog** [comment]
Hmm, then the title is not clear at all. You mean the more general https://github.com/in-toto/attestation?

**Comment by joshuagl** [comment]
> Oh I think some proposals would be great. Is the idea that everyone that wants to verify DSSE sigs would use in-toto to verify the roots?

I was thinking more along the lines of using TUF PKI-like features to distribute and manage public keys. 
Admittedly, if you are not using any of the other features of TUF it could be considered overkill, but it's serves a dual purpose for me as I get to write about TUF _and_ SLSA. :-)

**Comment by trishankatdatadog** [comment]
> I was thinking more along the lines of using TUF PKI-like features to distribute and manage public keys.
> Admittedly, if you are not using any of the other features of TUF it could be considered overkill, but it's serves a dual purpose for me as I get to write about TUF _and_ SLSA. :-)

We certainly can use TUF to verify in-toto attestations. However, what's not clear to me is whether we want a general way to cover _arbitrary_ DSSE payloads... it still can work, but I'm just not clear about what we're trying to do here.

**Comment by TomHennen** [comment]
I would definitely like a way to cover arbitrary DSSE payloads.

**Comment by trishankatdatadog** [comment]
> I would definitely like a way to cover arbitrary DSSE payloads.

It should work. @joshuagl, should we draft something up and send a PR?

**Comment by joshuagl** [comment]
> > I would definitely like a way to cover arbitrary DSSE payloads.
> 
> It should work. @joshuagl, should we draft something up and send a PR?

Absolutely, let's write something up (though I am AFK next week – let's catch up after?).

**Comment by trishankatdatadog** [comment]
@joshuagl let's get started on this?
